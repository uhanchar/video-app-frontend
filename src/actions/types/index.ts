export * from 'actions/types/app-types';
export * from 'actions/types/video-types';
