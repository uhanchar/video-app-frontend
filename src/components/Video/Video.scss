@import "src/constants";

.video {
  flex-direction: column;
  align-items: center;
  flex-grow: 1;
  height: calc(100vh - #{ $header-height });
  padding: 1.875rem 1.25rem;
  box-sizing: border-box;

  > .container {
    margin: 0 auto;

    > .item {
      position: relative;
      display: flex;
      justify-content: center;
      padding-top: 56.25%;

      > .player {
        position: absolute;
        top: 0;
        left: 0;
        width: 100% !important;
        height: 100% !important;

        > video {
          background: $primary-color-black;
        }
      }
    }

    > .text {
      display: flex;
      justify-content: space-between;

      > .text-content {
        text-align: justify;
      }
    }
  }
}
