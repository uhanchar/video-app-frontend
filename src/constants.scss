// colors
$color-blank: #ffffff;
$color-success: #2ed47a;
$primary-color-black: #020003;

// sizes
$header-height: 4rem;
