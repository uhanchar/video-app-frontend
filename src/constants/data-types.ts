export type NullableType<T> = T | null;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomAny = any;
