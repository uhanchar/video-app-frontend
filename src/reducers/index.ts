import { combineReducers } from 'redux';
import AppReducer from 'reducers/app.reducer';

const rootReducer = combineReducers({
  app: AppReducer,
});

export default rootReducer;
