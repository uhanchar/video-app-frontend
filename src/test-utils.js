import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { applyMiddleware, createStore } from 'redux';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';

import rootReducer from './reducers';

const reducerInitialState = {};

function render(
  ui,
  {
    initialState = reducerInitialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  } = {},
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      // eslint-disable-next-line react/jsx-filename-extension
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';

export { render };
