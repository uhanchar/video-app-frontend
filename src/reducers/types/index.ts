import { NullableType } from 'constants/data-types';

export interface IVideoItem {
  id: string;
  name: string;
  description: string;
  link: string;
  thumbnailLink: string;
  createdAt: Date;
}

export interface IAppState {
  videos: IVideoItem[];
}

export interface IVideoState {
  video: NullableType<IVideoItem>;
  isLoading: boolean;
  error: string;
}

export type RootState = {
  app: IAppState;
  video: IVideoState;
};
