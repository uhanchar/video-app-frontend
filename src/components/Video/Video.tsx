import React, { FC, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ReactPlayer from 'react-player';
import { Grid, Typography } from '@material-ui/core';

import 'components/Video/Video.scss';
import { clearVideoState, loadSelectedVideo } from 'actions/video.actions';
import { RootState } from 'reducers/types';
import { formatValidDate } from 'helpers/date-format.helper';
import Loader from 'components/Loader/Loader';
import Error from 'components/Error/Error';

interface IVideo {
  match: {
    params: {
      id: string;
    };
  };
}

const Video: FC<IVideo> = (props) => {
  const { id } = props.match.params;
  const dispatch = useDispatch();
  const { video, isLoading, error } = useSelector((state: RootState) => state.video);

  useEffect(() => {
    dispatch(loadSelectedVideo(id));

    return () => {
      dispatch(clearVideoState());
    };
  }, []);

  const renderLoader = () => (
    isLoading && (
      <Loader size="3rem" />
    )
  );

  const renderError = () => (
    error && (
      <Error content={error} />
    )
  );

  return (
    <div className="video">
      { renderLoader() }

      { renderError() }

      { !isLoading && video && !error && (
        <Grid container item className="container" justify="center" spacing={2} xs={12} sm={12} md={6}>
          <Grid item xs={12} className="item">
            <ReactPlayer url={`/${ video.link }`} controls className="player" />
          </Grid>

          <Grid item xs={12} className="text">
            <Typography variant="h4">{ video.name }</Typography>

            <Typography variant="overline" display="block">{ formatValidDate(video.createdAt) }</Typography>
          </Grid>

          <Grid item xs={12} className="text">
            <Typography variant="body2" gutterBottom className="text-content">{ video.description }</Typography>
          </Grid>
        </Grid>
      )}
    </div>
  );
};

export default Video;
