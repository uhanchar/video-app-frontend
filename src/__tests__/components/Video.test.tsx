import React from 'react';
import { applyMiddleware, createStore } from 'redux';
import axios from 'axios';
import { waitFor } from '@testing-library/react';
import thunk from 'redux-thunk';

import rootReducer from 'reducers';
import { render as renderWithRedux } from 'test-utils';
import Video from 'components/Video/Video';
import { IVideoItem, RootState } from 'reducers/types';
import { CustomAny } from 'constants/data-types';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

const videoItem: IVideoItem = {
  id: '39432ff1585cfb0ed1803ccea2929341',
  name: 'enem2',
  description: '',
  link: 'media/videos/39432ff1585cfb0ed1803ccea2929341',
  thumbnailLink: 'media/thumbnails/39432ff1585cfb0ed1803ccea2929341.png',
  createdAt: new Date(),
};

const initialState: RootState = {
  video: {
    video: null,
    isLoading: false,
    error: '',
  },
  app: {
    videos: [],
  },
};

const videoProps = {
  match: {
    params: {
      id: videoItem.id,
    },
  },
};

const errorMessage = 'Test error message';

let store: CustomAny;

describe('test for Video component', () => {
  beforeEach(() => {
    store = createStore(rootReducer, initialState, applyMiddleware(thunk));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render Video component', () => {
    const { asFragment } = renderWithRedux(<Video {...videoProps} />, { initialState, store });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render loader', () => {
    const { getAllByTestId } = renderWithRedux(<Video {...videoProps} />, { initialState, store });

    expect(getAllByTestId('loader').length).toBe(1);
  });

  it('should render video item name', async () => {
    await mockedAxios.get.mockImplementationOnce(() => Promise.resolve(videoItem));
    const { getByTestId } = renderWithRedux(<Video {...videoProps} />, { initialState, store });

    await waitFor(() => getByTestId('video-name'));

    expect(getByTestId('video-name')).toHaveTextContent(videoItem.name);
  });

  it('should render video item error', async () => {
    await mockedAxios.get.mockImplementationOnce(() => Promise.reject(errorMessage));
    const { getByTestId } = renderWithRedux(<Video {...videoProps} />, { initialState, store });

    await waitFor(() => getByTestId('error'));

    expect(getByTestId('error')).toHaveTextContent(errorMessage);
  });
});
