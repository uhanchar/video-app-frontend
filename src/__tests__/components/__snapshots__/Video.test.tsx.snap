// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for Video component should render Video component 1`] = `
<DocumentFragment>
  <div
    class="video"
  >
    <div
      class="loader-wrapper"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        data-testid="loader"
        role="progressbar"
        style="width: 3rem; height: 3rem;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
